core/storage/trie/impl/persistent_trie_batch_impl.cpp

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "storage/trie/impl/persistent_trie_batch_impl.hpp"
+
+#include "scale/scale.hpp"
+#include "storage/trie/impl/trie_error.hpp"
+
+namespace kagome::storage::trie {
+
+  const common::Buffer EXTRINSIC_INDEX_KEY{[]() {
+    return common::Buffer{}.put(":extrinsic_index");
+  }()};
+
+  // sometimes there is no extrinsic index for a runtime call
+  const common::Buffer NO_EXTRINSIC_INDEX_VALUE{[]() {
+    auto res = scale::encode(0xffffffff);
+    BOOST_ASSERT(res);
+    return res.value();
+  }()};
+
+  PersistentTrieBatchImpl::PersistentTrieBatchImpl(
+      std::shared_ptr<Codec> codec,
+      std::shared_ptr<TrieSerializer> serializer,
+      boost::optional<std::shared_ptr<changes_trie::ChangesTracker>> changes,
+      std::unique_ptr<PolkadotTrie> trie,
+      RootChangedEventHandler handler)
+      : codec_{std::move(codec)},
+        serializer_{std::move(serializer)},
+        changes_{std::move(changes)},
+        trie_{std::move(trie)},
+        root_changed_handler_{std::move(handler)} {
+    BOOST_ASSERT(codec_ != nullptr);
+    BOOST_ASSERT(serializer_ != nullptr);
+    BOOST_ASSERT((changes_.has_value() && changes_.value() != nullptr)
+                 or not changes_.has_value());
+    BOOST_ASSERT(trie_ != nullptr);
+    if (changes_) {
+      changes_.value()->setExtrinsicIdxGetter(
+          [this]() -> outcome::result<Buffer> {
+            auto res = get(EXTRINSIC_INDEX_KEY);
+            if (res.has_error() and res.error() == TrieError::NO_VALUE) {
+              return NO_EXTRINSIC_INDEX_VALUE;
+            }
+            return res;
+          });
```

Comment:
Because extrinsic index changes with every extrinsic, and to avoid referencing trie in changes tracker I pass lambda that allows it to obtain the latest extrinsic index.